File: Code/MestraTest/MestraGeneric/Trigger.cpp
#include "Trigger.h"

Trigger::Trigger()
{
}

bool Trigger::GetIsEnabled() const
{
	return (_data[0] & 0x80) == 0x80; // MSB
}

void Trigger::SetIsEnabled(bool enabled)
{
	_data[0] = static_cast<unsigned char>((_data[0] & 0x7F) + (enabled ? 0x80 : 0x00));
}
 

Trigger::EMidiTriggerType Trigger::GetMidiTriggerType() const
{
	return static_cast<EMidiTriggerType>(_data[0] & 0x0F);
}

void Trigger::SetMidiTriggerType(EMidiTriggerType midiTriggerType)
{
	_data[0] = static_cast<unsigned char>((_data[0] & 0x80) + static_cast<unsigned char >(midiTriggerType));   // MIDI = 000 (bits 6~4
}


int Trigger::GetMidiChannel() const
{
	return _data[1];
}

void Trigger::SetMidiChannel(int midiChannel)
{
	_data[1] = static_cast<unsigned char>(midiChannel);
}


bool Trigger::GetIsGroup() const
{
	return (_data[2] & 0x80) == 0x80;
}

void Trigger::SetIsGroup(bool isGroup)
{
	_data[2] = static_cast<unsigned char>((_data[2] & 0x7F) + (isGroup ? 0x80 : 0x00));
}


bool Trigger::GetAll() const
{
	return (_data[2] & 0x40) == 0x40;
}

void Trigger::SetAll(bool all)
{
	_data[2] = static_cast<unsigned char>((_data[2] & 0xBF) + (all ? 0x40 : 0x00));
}


int Trigger::GetGroup() const
{
	return (_data[2] & 0x3C) >> 2;
}

void Trigger::SetGroup(int group)
{
	_data[2] = static_cast<unsigned char>((_data[2] & 0xC3) + (group << 2));
}


int Trigger::GetValue() const
{
	return _data[2] & 0x7F;
}

void Trigger::SetValue(int value)
{
	_data[2] = static_cast<unsigned char>((_data[2] & 0x80) + value);
}

/// <summary>
/// Copy data.
/// </summary>
/// <returns></returns>
void Trigger::Clone(Trigger* trigger)
{
	this[0] = trigger[0];
	this[1] = trigger[1];
	this[2] = trigger[2];
}

File: Code/MestraTest/MestraGeneric/Trigger.cs
using System;

namespace MestraGeneric
{
    public class Trigger : ICloneable
    {
        public byte[] Data  { get; set; }

        public enum EMidiTriggerType
        {
            NoteOn = 0x8,
            NoteOff = 0x9
        }

        public Trigger()
        {
            Data = new byte[3];
        }

        public bool IsEnabled
        {
            get { return (Data[0] & 0x80) == 0x80; } // MSB
            set { Data[0] = (byte)((Data[0] & 0x7F) + (value ? 0x80 : 0x00)); }
        }

        public EMidiTriggerType MidiTriggerType
        {
            get { return (EMidiTriggerType)(Data[0] & 0x0F); }
            set { Data[0] = (byte)((Data[0] & 0x80) + (byte)value); } // MIDI = 000 (bits 6~4
        }

        public int MidiChannel
        {
            get { return Data[1]; }
            set { Data[1] = (byte)value; }
        }

        public bool IsGroup
        {
            get { return (Data[2] & 0x80) == 0x80; }
            set { Data[2] = (byte)((Data[2] & 0x7F) + (value ? 0x80 : 0x00)); }
        }

        public bool All
        {
            get { return (Data[2] & 0x40) == 0x40; }
            set { Data[2] = (byte)((Data[2] & 0xBF) + (value ? 0x40 : 0x00)); }
        }

        public int Group
        {
            get { return (Data[2] & 0x3C) >> 2; }
            set { Data[2] = (byte)((Data[2] & 0xC3) + (value << 2)); }
        }


        public int Value
        {
            get { return Data[2] & 0x7F; }
            set { Data[2] = (byte)((Data[2] & 0x80) + value); }
        }

        /// <summary>
        /// Copy data.
        /// </summary>
        /// <returns></returns>
        public object Clone()
        {
            var trigger = (Trigger)MemberwiseClone();
            trigger.Data = new[] { Data[0], Data[1], Data[2] };
            return trigger;
        }
    }
}
